src/v3/tutorial/middleware/README.md

```diff
@@ -0,0 +1,87 @@
+---
+description: This tutorial will show you how to use a middleware.
+full_title: Middleware Tutorial
+---
+
+# Middleware Tutorial
+
+This tutorial will show you how to use a middleware. A middleware is used when an application wants to provide a way for other applications to change its behavior.
+
+## Usage
+
+```javascript
+// To add a callback to a middleware group
+core.make('osjs/middleware').add('middleware-group-name', middlewareCallback);
+// OR
+import osjs from 'osjs';
+
+osjs.middleware('middleware-group-name', middlewareCallback);
+
+// To get the list of callbacks in a middleware group
+const middlewareCallbacks = core.make('osjs/middleware').get('middleware-group-name');
+```
+
+## Example
+
+In an application like a virus scanner we need to add a "Scan" option in the context menu of files in the [File Manager application](https://github.com/os-js/osjs-filemanager-application) that opens the scanner application.
+
+![Example](example.png)
+
+The [File Manager application](https://github.com/os-js/osjs-filemanager-application) should
+provide a way for other applications to add to its edit/context menu. So, it uses a middleware group for its edit/context menu.
+```javascript
+// File Manager when rendering edit/context menu
+const items = core.make('osjs/middleware').get('osjs/filemanager:menu:edit');
+
+const middlewarePromises = items.map(fn => fn({file, isContextMenu}));
+const middlewareValues = await Promise.all(middlewarePromises);
+
+const middlewareItems = middlewareValues.filter(items => items instanceof Array);
+
+// It adds the items in the `middlewareItems` element to the context/edit menu
+```
```

Comment:
Hm. I'm not sure if it is a good idea to show the implementation in the FileManager here, for a couple of reasons:

1. If the FM code changes, this also have to be updated
2. Only a portion of the implementation is showed which is basically the same as the example in usage section
3. It can be a bit overwhelming or confusing that this is here

So if you agree with my suggestion of generalizing the example in the comment above this section could just say:

```
The middleware group name the File Manager exposes for this is `osjs/filemanager:menu:edit`.
```